traits/observers/_list_item_observer.py

```diff
@@ -0,0 +1,221 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+from traits.observers._list_change_event import list_event_factory
+from traits.observers._i_observer import IObserver
+from traits.observers._trait_event_notifier import TraitEventNotifier
+from traits.observers._observe import add_or_remove_notifiers
+from traits.observers._observer_change_notifier import ObserverChangeNotifier
+from traits.trait_list_object import TraitList
+
+
+@IObserver.register
+class ListItemObserver:
+    """ Observer for observing mutations on a list.
+
+    Parameters
+    ----------
+    notify : boolean
+        Whether to notify for changes.
+    optional : boolean
+        If False, this observer will complain if the incoming object is not
+        an observable list. If True and the incoming object is not a list, this
+        observer will do nothing. Useful for the 'items' keyword in the text
+        parser, where the source container type is ambiguous.
+    """
+
+    def __init__(self, *, notify, optional):
+        self._notify = notify
+        self._optional = optional
+
+    def __hash__(self):
+        """ Return a hash of this object."""
+        return hash((type(self), self.notify, self.optional))
```

Comment:
The use of `type(self)` here makes me a bit uncomfortable: it introduces an extra level of non-deterministic behaviour, since memory address, and hence the hash, of the type will likely vary from run to run. How about using the type's name instead? That at least gives us the option of making things determinstic (e.g., during debugging) by setting `PYTHONHASHSEED`.

Reply:
Initially I prefer not using the name because it assumes two classes must have unique name, but it is not true, e.g. There are two `TraitList` classes, and two `Expression` classes in the traits distribution.

We could include the fully qualified module names though. I think that's what I am going to do.

Reply:
> it assumes two classes must have unique name

Why? There's no rule that says that hashes have to be unique - and in fact, they can't be. For practical purposes, all you want is to ensure that for unequal things that are likely to end up in a set or dict together, it's not too likely for those things to have the same hash value. And if they do, everything will still work just fine; it might just be a tiny amount slower thanks to some extra collisions internally.

So it's fine to just use the type name here; there's no need for fully qualified names. (Nothing _wrong_ with using fully qualified names if that's what you want to do, but it doesn't bring any benefit.)

Reply:
That's true.